File: 0115-distinct-subsequences/0115-distinct-subsequences.cpp
class Solution {
public:

    int func(string s, string t, vector<vector<int>>&dp, int i, int j){
        if(j==t.size()) return 1;
        if(i==s.size()) return 0;
        if(dp[i][j]!=-1) return dp[i][j];

        if(s[i]==t[j]) dp[i][j]=func(s,t,dp,i+1,j+1)+func(s,t,dp,i+1,j);
        else dp[i][j]=func(s,t,dp,i+1,j);
        return dp[i][j];

    }

    int numDistinct(string s, string t) {
        int n=s.size();
        int m=t.size();
        if(m>n) return 0;
        vector<vector<int>>dp(n+1,vector<int>(m+1,-1));
        return func(s,t,dp,0,0);

    }
};

File: 0115-distinct-subsequences/0115-distinct-subsequences.js
/**
 * @param {string} s
 * @param {string} t
 * @return {number}
 */
var numDistinct = function(s, t) {
    const m = t.length;

    const dp = new Array(m + 1).fill(0);
    dp[0] = 1;

    for (let i = 0; i < s.length; i++) {
        // Traverse backwards to avoid using an updated
        // dp[j - 1] from the same iteration.
        for (let j = m; j >= 1; j--) {
            if (s[i] === t[j - 1]) {
                dp[j] += dp[j - 1];
            }
        }
    }

    return dp[m];
};